discovery-provider/src/tasks/index_metrics.py

```diff
@@ -126,7 +136,154 @@ def refresh_metrics_matviews(db):
+    logger.info(f"number of discovery providers: {num_discovery_providers}")
+    service_infos = [sp_factory_inst.functions.getServiceEndpointInfo(discovery_node_service_type, i).call() \
+        for i in range(1, num_discovery_providers + 1)]
+
+    logger.info(f"this node's delegate owner wallet: {shared_config['delegate']['owner_wallet']}")
+    all_other_nodes = [service_info[1] for service_info in service_infos \
+        if service_info[3] != shared_config["delegate"]["owner_wallet"]]
+    logger.info(f"all the other nodes: {all_other_nodes}")
+    return all_other_nodes
+
+def get_metrics(endpoint, start_time):
+    try:
+        route_metrics_endpoint = f"{endpoint}/v1/metrics/routes/cached?start_time={start_time}"
+        logger.info(f"route metrics request to: {route_metrics_endpoint}")
+        route_metrics_response = requests.get(route_metrics_endpoint, timeout=10)
+        if route_metrics_response.status_code != 200:
+            raise Exception(f"Query to cached route metrics endpoint {route_metrics_endpoint} \
+                failed with status code {route_metrics_response.status_code}")
+
+        app_metrics_endpoint = f"{endpoint}/v1/metrics/apps/cached?start_time={start_time}"
+        logger.info(f"app metrics request to: {app_metrics_endpoint}")
+        app_metrics_response = requests.get(app_metrics_endpoint, timeout=10)
+        if app_metrics_response.status_code != 200:
+            raise Exception(f"Query to cached app metrics endpoint {app_metrics_endpoint} \
+                failed with status code {app_metrics_response.status_code}")
+
+        return route_metrics_response.json()['data'], app_metrics_response.json()['data']
+    except Exception as e:
+        logger.warning(f"Could not get metrics from node {endpoint} with start_time {start_time}")
+        logger.error(e)
+        return None, None
+
+"""
+Get recent route and app metrics from all other discovery nodes
+and merge with this node's metrics so that this node will be aware
+of all the metrics across users hitting different providers
+"""
+def consolidate_metrics_from_other_nodes(self, db, redis):
+    all_other_nodes = get_all_other_nodes()
+
+    visited_node_timestamps_str = redis_get_or_restore(redis, metrics_visited_nodes)
+    visited_node_timestamps = json.loads(visited_node_timestamps_str) if visited_node_timestamps_str else {}
+
+    one_iteration_ago = datetime.utcnow() - timedelta(minutes=METRICS_INTERVAL)
+    one_iteration_ago_str = one_iteration_ago.strftime(datetime_format_secondary)
+    for node in all_other_nodes:
+        start_time_str = visited_node_timestamps[node] if node in visited_node_timestamps else one_iteration_ago_str
+        start_time_obj = datetime.strptime(start_time_str, datetime_format_secondary)
+        start_time = int(start_time_obj.timestamp())
+        new_route_metrics, new_app_metrics = get_metrics(node, start_time)
+
+        logger.info(f"received route metrics: {new_route_metrics}")
```

Comment:
this will probably be a massive object to print, do we want to output this?

Reply:
good point